lib/pool.dart

```diff
@@ -121,6 +121,44 @@ class Pool {
     });
   }
 
+  /// Requests a resource for the combined duration of the [callback] and the
+  /// processing of the Stream it returns.
+  /// (E.g. a database connection is hold until the the results get processed.)
+  Stream/*<T>*/ streamWithResource/*<T>*/(Stream/*<T>*/ callback()) {
+    StreamController/*<T>*/ streamController;
+    StreamSubscription/*<T>*/ streamSubscription;
+    Completer resourceCompleter = new Completer();
+    final complete = () {
+      if (streamSubscription != null) {
+        streamSubscription.cancel();
+        streamSubscription = null;
+      }
+      if (!resourceCompleter.isCompleted) {
+        resourceCompleter.complete();
+        streamController.close();
+      }
+    };
+    streamController = new StreamController/*<T>*/(onCancel: complete);
+    withResource(() {
+      if (resourceCompleter.isCompleted) return null;
+      new Future.sync(callback).then((Stream/*<T>*/ stream) {
+        if (stream == null) {
+          complete();
+          return null;
+        }
+        streamSubscription = stream.listen(streamController.add,
+            onError: streamController.addError,
+            onDone: complete,
+            cancelOnError: true);
```

Comment:
> If you add .listen(null) it works.

Not sure how this should work. The returned Future of `withResource` will complete, it doesn't matter if we add an `await` or `.then()` on it. I think the same should be true for Streams (e.g. it should complete with or without a subscriber), but can be convinced otherwise.

> if you're willing to wait a day or two

sure, thanks for jumping on this!

> Can't you just use the returned stream's "done" event to check this?

How can I listen on Stream's done? The `drain` will introduce another listener, which will fail if it is single-subscription. `StreamController.done` is not available when all I got is a Stream.